### PR TITLE
 Avoid a branch on key being local for queries that use the same local and extern providers

### PR DESCRIPTION
**Pull request by bjorn3**
Currently based on https://github.com/rust-lang/rust/pull/85810 as it slightly conflicts with it. Only the last two commits are new.

### CODE REVIEW

**Comment by bjorn3** [comment]
On average this is an improvement, but there is also a non-trivial amount of regressions. Maybe different inlining?

File: compiler/rustc_middle/src/dep_graph/dep_node.rs [hunk 1]
```diff
@@ -140,11 +140,11 @@ macro_rules! is_eval_always_attr {
 }
 
 macro_rules! contains_anon_attr {
-    ($($attr:ident $(($($attr_args:tt)*))* ),*) => ({$(is_anon_attr!($attr) | )* false});
```
Comment (cjgillot) [inline compiler/rustc_middle/src/dep_graph/dep_node.rs:143]: Can all these macros be simplified now?

File: compiler/rustc_metadata/src/rmeta/decoder/cstore_impl.rs [hunk 2]
```diff
@@ -56,7 +56,7 @@ macro_rules! provide {
                 $compute
             })*
 
-            *providers = Providers {
+            *providers = ExternProviders {
```
Comment (cjgillot) [inline compiler/rustc_metadata/src/rmeta/decoder/cstore_impl.rs:59]: Is this provide exhaustive now? Should it be made so?

**Review by cjgillot — COMMENTED** [review]
I am not convinced yet that removing one branch is worth the added complexity. The branch itself should be optimized out.

Can we ensure statically that `Key::query_crate` can only return non-`LOCAL_CRATE` when `separate_provide_extern` is set? Should we even?

File: compiler/rustc_metadata/src/rmeta/decoder/cstore_impl.rs (hunk 2 above)
Comment (bjorn3) [reply compiler/rustc_metadata/src/rmeta/decoder/cstore_impl.rs:59]: Due to macro implementation issues I am using `()` as type for queries that don't have a separate extern provider.

**Comment by bjorn3** [comment]
> I am not convinced yet that removing one branch is worth the added complexity. The branch itself should be optimized out.

It has a side effect of for the most part documenting which queries are directly calculated from HIR and whose calculated value may be stored in the crate metadata and which are calculated in the current compilation session from other queries and configuration options.

**Comment by JohnCSimon** [comment]
@cjgillot Can you please resolve the merge conflict?

@rustbot -S-waiting-on-review +S-waiting-on-author

**Comment by JohnCSimon** [comment]
triage:
@bjorn3 -can you please resolve the merge conflict?

**Pull request merged**